File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/kakuro/domain/KakuroGameState.java
package com.zwstudio.logicpuzzlesandroid.puzzles.kakuro.domain;

import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGameState;
import com.zwstudio.logicpuzzlesandroid.common.domain.Graph;
import com.zwstudio.logicpuzzlesandroid.common.domain.MarkerOptions;
import com.zwstudio.logicpuzzlesandroid.common.domain.Node;
import com.zwstudio.logicpuzzlesandroid.common.domain.Position;
import com.zwstudio.logicpuzzlesandroid.home.domain.HintState;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import fj.F;

import static fj.data.HashMap.fromMap;
import static fj.data.List.iterableList;

/**
 * Created by zwvista on 2016/09/29.
 */

public class KakuroGameState extends CellsGameState<KakuroGame, KakuroGameMove, KakuroGameState> {
    public KakuroObject[] objArray;
    public HintState[] row2state;
    public HintState[] col2state;

    public KakuroGameState(KakuroGame game) {
        super(game);
        objArray = new KakuroObject[rows() * cols()];
        Arrays.fill(objArray, KakuroObject.Empty);
        row2state = new HintState[rows()];
        col2state = new HintState[cols()];
        updateIsSolved();
    }

    public KakuroObject get(int row, int col) {
        return objArray[row * cols() + col];
    }
    public KakuroObject get(Position p) {
        return get(p.row, p.col);
    }
    public void set(int row, int col, KakuroObject obj) {
        objArray[row * cols() + col] = obj;
    }
    public void set(Position p, KakuroObject obj) {
        set(p.row, p.col, obj);
    }

    public boolean setObject(KakuroGameMove move) {
        Position p = move.p;
        if (!isValid(p) || get(p) == move.obj) return false;
        set(p, move.obj);
        updateIsSolved();
        return true;
    }

    public boolean switchObject(KakuroGameMove move) {
        MarkerOptions markerOption = MarkerOptions.values()[game.gdi.getMarkerOption()];
        F<KakuroObject, KakuroObject> f = obj -> {
            switch (obj) {
            case Empty:
                return markerOption == MarkerOptions.MarkerFirst ?
                        KakuroObject.Marker : KakuroObject.Cloud;
            case Cloud:
                return markerOption == MarkerOptions.MarkerLast ?
                        KakuroObject.Marker : KakuroObject.Empty;
            case Marker:
                return markerOption == MarkerOptions.MarkerFirst ?
                        KakuroObject.Cloud : KakuroObject.Empty;
            }
            return obj;
        };
        Position p = move.p;
        if (!isValid(p)) return false;
        move.obj = f.f(get(p));
        return setObject(move);
    }

    /*
        iOS Game: Logic Games/Puzzle Set 5/Kakuro

        Summary
        Weather Radar Report

        Description
        1. You must find Kakuro in the sky.
        2. The hints on the borders tell you how many tiles are covered by Kakuro
           in that row or column.
        3. Kakuro only appear in rectangular or square areas. Furthermore, their
           width and height is always at least two tiles wide.
        4. Kakuro can't touch between themselves, not even diagonally. 
    */
    private void updateIsSolved() {
        boolean allowedObjectsOnly = game.gdi.isAllowedObjectsOnly();
        isSolved = true;
        for (int r = 0; r < rows(); r++)
            for (int c = 0; c < cols(); c++)
                if (get(r, c) == KakuroObject.Forbidden)
                    set(r, c, KakuroObject.Empty);
        for (int r = 0; r < rows(); r++) {
            int n1 = 0, n2 = game.row2hint[r];
            for (int c = 0; c < cols(); c++)
                if (get(r, c) == KakuroObject.Cloud)
                    n1++;
            row2state[r] = n1 < n2 ? HintState.Normal : n1 == n2 ? HintState.Complete : HintState.Error;
            if (n1 != n2) isSolved = false;
        }
        for (int c = 0; c < cols(); c++) {
            int n1 = 0, n2 = game.col2hint[c];
            for (int r = 0; r < rows(); r++)
                if (get(r, c) == KakuroObject.Cloud)
                    n1++;
            col2state[c] = n1 < n2 ? HintState.Normal : n1 == n2 ? HintState.Complete : HintState.Error;
            if (n1 != n2) isSolved = false;
        }
        for (int r = 0; r < rows(); r++)
            for (int c = 0; c < cols(); c++) {
                KakuroObject o = get(r, c);
                if ((o == KakuroObject.Empty || o == KakuroObject.Marker) && allowedObjectsOnly && (
                        row2state[r] != HintState.Normal || col2state[c] != HintState.Normal))
                    set(r, c, KakuroObject.Forbidden);
            }
        if (!isSolved) return;
        Graph g = new Graph();
        Map<Position, Node> pos2node = new HashMap<>();
        for (int r = 0; r < rows(); r++)
            for (int c = 0; c < cols(); c++) {
                Position p = new Position(r, c);
                if (get(p) != KakuroObject.Cloud) continue;
                Node node = new Node(p.toString());
                g.addNode(node);
                pos2node.put(p, node);
            }
        for (Position p : pos2node.keySet())
            for (Position os : KakuroGame.offset) {
                Position p2 = p.add(os);
                if (pos2node.containsKey(p2))
                    g.connectNode(pos2node.get(p), pos2node.get(p2));
            }
        while (!pos2node.isEmpty()) {
            g.setRootNode(iterableList(pos2node.values()).head());
            List<Node> nodeList = g.bfs();
            int r2 = 0, r1 = rows(), c2 = 0, c1 = cols();
            for (Node node : nodeList) {
                Position p = fromMap(pos2node).toStream().find(e -> e._2().equals(node)).some()._1();
                pos2node.remove(p);
                if (r2 < p.row) r2 = p.row;
                if (r1 > p.row) r1 = p.row;
                if (c2 < p.col) c2 = p.col;
                if (c1 > p.col) c1 = p.col;
            }
            int rs = r2 - r1 + 1, cs = c2 - c1 + 1;
            if (!(rs >= 2 && cs >= 2 && rs * cs == nodeList.size())) {
                isSolved = false;
                return;
            }
        }
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/kakuro/domain/KakuroGameState.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.kakuro.domain

import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGameState
import com.zwstudio.logicpuzzlesandroid.common.domain.Position
import com.zwstudio.logicpuzzlesandroid.home.domain.HintState
import java.util.*

class KakuroGameState(game: KakuroGame) : CellsGameState<KakuroGame?, KakuroGameMove?, KakuroGameState?>(game) {
    var pos2num: MutableMap<Position?, Int?> = HashMap()
    var pos2horzHint: MutableMap<Position?, HintState?> = HashMap()
    var pos2vertHint: MutableMap<Position?, HintState?> = HashMap()
    operator fun get(p: Position?) = pos2num[p]!!

    operator fun set(p: Position?, obj: Int?) {
        pos2num[p] = obj
    }

    fun setObject(move: KakuroGameMove?): Boolean {
        val p = move!!.p
        if (!isValid(p) || get(p) == null || get(p) == move.obj) return false
        set(p, move.obj)
        updateIsSolved()
        return true
    }

    fun switchObject(move: KakuroGameMove?): Boolean {
        val p = move!!.p
        if (!isValid(p) || get(p) == null) return false
        val o = get(p)
        move.obj = (o + 1) % 10
        return setObject(move)
    }

    /*
        iOS Game: Logic Games/Puzzle Set 4/Kakuro

        Summary
        Fill the board with numbers 1 to 9 according to the sums

        Description
        1. Your goal is to write a number in every blank tile (without a diagonal
           line).
        2. The number on the top of a column or at the left of a row, gives you
           the sum of the numbers in that column or row.
        3. You can write numbers 1 to 9 in the tiles, however no same number should
           appear in a consecutive row or column.
        4. Tiles which only have a diagonal line aren't used in the game.
    */
    private fun updateIsSolved() {
        isSolved = true
        for ((p, n2) in game!!.pos2horzHint) {
            var n1 = 0
            val os: Position = KakuroGame.Companion.offset.get(1)
            var n: Int
            val lastN = 0
            val p2 = p.add(os)
            while (pos2num[p2].also { n = it!! } != null) {
                n1 += n
                // 3. You can write numbers 1 to 9 in the tiles, however no same number should
                // appear in a consecutive row.
                if (n == lastN) {
                    isSolved = false
                    pos2horzHint[p2] = HintState.Error
                    pos2horzHint[p2.subtract(os)] = HintState.Error
                }
                p2.addBy(os)
            }
            // 2. The number on at the left of a row gives you
            // the sum of the numbers in that row.
            val s = if (n1 == 0) HintState.Normal else if (n1 == n2) HintState.Complete else HintState.Error
            pos2horzHint[p] = s
            if (s != HintState.Complete) isSolved = false
        }
        for ((p, n2) in game!!.pos2vertHint) {
            var n1 = 0
            val os: Position = KakuroGame.Companion.offset.get(2)
            var n: Int
            val lastN = 0
            val p2 = p.add(os)
            while (pos2num[p2].also { n = it!! } != null) {
                n1 += n
                // 3. You can write numbers 1 to 9 in the tiles, however no same number should
                // appear in a consecutive column.
                if (n == lastN) {
                    isSolved = false
                    pos2horzHint[p2] = HintState.Error
                    pos2horzHint[p2.subtract(os)] = HintState.Error
                }
                p2.addBy(os)
            }
            // 2. The number on the top of a column gives you
            // the sum of the numbers in that column.
            val s = if (n1 == 0) HintState.Normal else if (n1 == n2) HintState.Complete else HintState.Error
            pos2vertHint[p] = s
            if (s != HintState.Complete) isSolved = false
        }
    }

    init {
        pos2num = HashMap(game.pos2num)
        updateIsSolved()
    }
}